core/arch/interp.c

```diff
@@ -2836,11 +2836,17 @@ client_process_bb(dcontext_t *dcontext, build_bb_t *bb)
 #    ifdef X86
         if (!d_r_is_avx512_code_in_use()) {
             if (ZMM_ENABLED()) {
-                if (instr_may_write_zmm_register(inst)) {
-                    LOG(THREAD, LOG_INTERP, 2, "Detected AVX-512 code in use\n");
-                    d_r_set_avx512_code_in_use(true);
-                    proc_set_num_simd_saved(MCXT_NUM_SIMD_SLOTS);
+#        ifdef CLIENT_INTERFACE
+                if (!TEST(INSTR_CLEANCALL_SIMD_CONTEXT, inst->flags)) {
```

Comment:
This should look only at app instructions in the first place.  This flag is not necessary.

Reply:
It also looks at client instructions. If the client instruments with AVX-512, the lazy switch (should-) kick(s) in. The flag is there in order to distinguish context saving vmovs from instrumented code. Do you have a better idea than using the flag to differentiate the two?

Reply:
> If the client instruments with AVX-512, the lazy switch (should-) kick(s) in. 

Why?  If the app is not using AVX-512 there is no reason to swap the context.

Reply:
In https://github.com/DynamoRIO/dynamorio/pull/3738 I added client AVX-512 code detection. The reason there was that a client may be compiled with AVX-512.

Reply:
Where specifically in #3738?  If a client is compiled with AVX-512 how are we going to detect it?  It does not seem feasible.  Without some help from the client telling us whether it uses AVX-512 I don't see how we can do anything but assume it might use it.  We should revisit the lazy design: I think it only matters on attach.  Please think it through and update the doc.

Reply:
Offline. Fixed using !instr_is_meta(). I am updating the doc now. A missing part that I am adding is that there will be a client API for the client to tell DynamoRIO that it is compiled with AVX-512. We should add to the documentation that clients should not be compiled with AVX-512 if possible, and if they are, they should call this function, because DynamoRIO can't detect whether they are clobbering AVX-512 state.

Reply:
(Note that instr_is_app is certainly preferred in client code (see tutorial and docs: using "app" and "tool" labels and moving away from the sometimes-confusing "meta"))

Your goal by adding this new API is to keep lazy cxt switching for attach?

Reply:
The goal for this client function would be to disable laziness if the client itself is clobbering AVX-512 state, and to turn on preserving AVX-512 state right from the start (would need to be called at init time).

Reply:
But if the app is not using 512, and DR is in control from the start, it doesn't matter whether the client clobbers 512, does it?  Whenever the app starts using it, it will have to copy in new state from somewhere else, right?  It seems to me this only matters for attach, where there can be app state in place before DR takes over.  In that case, the client clobbering it matters and prevents DR being lazy.

Reply:
I think this is true. The new function could say in its commented that a client only needs to call it if used for attach. Might be simpler - and safer if a client would always call it? You don't know after all which case a client library is used for?

Reply:
The new function should enable me to only query the lazy switch in https://github.com/DynamoRIO/dynamorio/pull/3751, leaving it to the client to call the new API function. Looking at the cases we're calling PUSH_PRIV_MCXT currently, the attach ones are the ones that matter (I guess as you mentioned above)

Reply:
Since this is a build feature, instead of a function I would make a CMake option which sets (or removes) the needed compiler flags and sets a var in the library like the current version var or sthg.

Reply:
That's a good idea. Really neat would be if there was code that every client needs to include (configure.h) and set this based on __AVX512F__?

Reply:
Yes, that is how the version var is set.